src/libaktualizr/primary/sotauptaneclient.cc

```diff
@@ -1162,43 +1211,40 @@ std::vector<result::Install::EcuReport> SotaUptaneClient::sendImagesToEcus(const
-        const std::string fw = sstr.str();
-        firmwareFutures.emplace_back(result::Install::EcuReport(*targets_it, ecu_serial, data::InstallationResult()),
-                                     sendFirmwareAsync(sec, std::make_shared<std::string>(fw)));
-      }
+      Uptane::SecondaryInterface &sec = *f->second;
+      firmwareFutures.emplace_back(result::Install::EcuReport(*targets_it, ecu_serial, data::InstallationResult()),
+                                   sendFirmwareAsync(sec, *targets_it));
     }
   }
 
   for (auto &f : firmwareFutures) {
-    // failure
+    data::ResultCode::Numeric fut_result = data::ResultCode::Numeric::kUnknown;
+
+    fut_result = f.second.get();
+    f.first.install_res = data::InstallationResult(fut_result, "");
+
     if (fiu_fail((std::string("secondary_install_") + f.first.serial.ToString()).c_str()) != 0) {
-      f.first.install_res = data::InstallationResult(
-          data::ResultCode(data::ResultCode::Numeric::kInstallFailed, fault_injection_last_info()), "");
-      storage->saveEcuInstallationResult(f.first.serial, f.first.install_res);
-      reports.push_back(f.first);
-      continue;
+      // consider changing this approach of the fault injection, since the current approach impacts the non-test code
+      // flow here as well as it doesn't test the installation failure on secondary from an end-to-end perspective as it
+      // injects an error on the middle of the control flow that would have happened if an installation error had
+      // happened in case of the virtual or the ip-secondary or any other secondary, e.g. add a mock secondary that
+      // returns an error to sendFirmware/install request we might consider passing the installation description message
+      // from Secondary, not just bool and/or data::ResultCode::Numeric
+
+      // fault injection
+      fut_result = data::ResultCode::Numeric::kInstallFailed;
+      f.first.install_res = data::InstallationResult(data::ResultCode(fut_result, fault_injection_last_info()), "");
     }
 
-    bool fut_result = f.second.get();
-    if (fut_result) {
-      f.first.install_res = data::InstallationResult(data::ResultCode::Numeric::kOk, "");
-      storage->saveInstalledVersion(f.first.serial.ToString(), f.first.update, InstalledVersionUpdateMode::kCurrent);
-    } else {
-      f.first.install_res = data::InstallationResult(data::ResultCode::Numeric::kInstallFailed, "");
+    if (fut_result == data::ResultCode::Numeric::kOk || fut_result == data::ResultCode::Numeric::kNeedCompletion) {
+      f.first.update.setCorrelationId(director_repo.getCorrelationId());
+      storage->saveInstalledVersion(f.first.serial.ToString(), f.first.update,
+                                    fut_result == data::ResultCode::Numeric::kOk
+                                        ? InstalledVersionUpdateMode::kCurrent
+                                        : InstalledVersionUpdateMode::kPending);
```

Comment:
Not a big deal or blocker here, but I'd probably prefer to have the three-line logic for calculating the result on its own line outside of the function call.

Reply:
ok